system-parachains/asset-hubs/asset-hub-polkadot/src/ah_migration/xcm_mapping.rs

```diff
@@ -0,0 +1,196 @@
+	ah_location: &Location,
+	universal_location: &InteriorLocation,
+) -> Result<Instruction<()>, ()> {
+	use Instruction::*;
+
+	// We only map a particular subset of instructions, to keep the logic minimal.
+	match instruction {
+		WithdrawAsset(assets) => {
+			let reanchored_assets =
+				assets.reanchored(ah_location, universal_location).map_err(|err| {
+					log::error!(target: LOG_TARGET, "Failed to reanchor assets: {err:?}");
+				})?;
+			Ok(WithdrawAsset(reanchored_assets))
+		},
+		PayFees { asset } => {
+			let reanchored_asset =
+				asset.reanchored(ah_location, universal_location).map_err(|err| {
+					log::error!(target: LOG_TARGET, "Failed to reanchor asset: {err:?}");
+				})?;
+			Ok(PayFees { asset: reanchored_asset })
+		},
+		DepositAsset { assets, beneficiary } => {
+			let reanchored_beneficiary = if is_local_account(&beneficiary) {
+				// Local accounts (parents: 0, AccountId32/AccountKey20) don't need reanchoring.
+				beneficiary
+			} else {
+				beneficiary.reanchored(ah_location, universal_location).map_err(|err| {
+					log::error!(target: LOG_TARGET, "Failed to reanchor beneficiary: {err:?}");
+				})?
+			};
+			Ok(DepositAsset { assets, beneficiary: reanchored_beneficiary })
+		},
+		DepositReserveAsset { assets, dest, xcm } => {
+			let reanchored_dest =
+				dest.reanchored(ah_location, universal_location).map_err(|err| {
+					log::error!(target: LOG_TARGET, "Failed to reanchor dest: {err:?}");
+				})?;
+			// The nested xcm is already from the perspective of the dest, so no reanchoring needed.
+			Ok(DepositReserveAsset { assets, dest: reanchored_dest, xcm })
+		},
+		RefundSurplus => Ok(RefundSurplus),
+
+		// All other instructions are not supported in the migration.
+		instruction => {
+			log::error!(
+				target: LOG_TARGET,
+				"Unsupported XCM instruction in migration: {instruction:?}",
+			);
+			Err(())
+		},
+	}
+}
+
+fn is_local_account(location: &Location) -> bool {
+	// Check if this is a local account (parents: 0, AccountId32/AccountKey20)
+	if location.parents != 0 {
+		return false;
+	}
+
+	match location.interior.first() {
+		Some(Junction::AccountId32 { .. }) | Some(Junction::AccountKey20 { .. }) => true,
+		_ => false,
+	}
+}
+
+fn convert_descend_to_alias_origin(instruction: Instruction<()>) -> Result<Instruction<()>, ()> {
+	use Instruction::*;
+
+	match instruction {
+		DescendOrigin(interior) => {
+			// Convert `DescendOrigin` to `AliasOrigin` from Asset Hub.
+			// `DescendOrigin` by the Relay Chain operated from one parent up,
+			// so we need to add one parent	to the location when converting to `AliasOrigin`.
+			let alias_location = Location::new(1, interior);
+			Ok(AliasOrigin(alias_location))
+		},
+		// All other instructions pass through unchanged
+		instruction => Ok(instruction),
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use crate::ah_migration::{RcRuntimeCall, RcToAhCall, RcXcmCall, RuntimeCall};
+	use xcm::prelude::*;
+
+	#[test]
+	fn map_xcm_execute() {
```

Comment:
I think this ref could be a good test. You can load the raw data through PJS and then just past it as Hex string in here and decode: https://polkadot.subsquare.io/referenda/1770